src/NuGet.Core/NuGet.Commands/TrustedSignersCommand/TrustedSignerActionsProvider.cs

```diff
@@ -0,0 +1,310 @@
+                throw new ArgumentException(Strings.ArgumentCannotBeNullOrEmpty, nameof(name));
+            }
+
+            if (!Enum.IsDefined(typeof(VerificationTarget), trustTarget) || trustTarget == VerificationTarget.Unknown)
+            {
+                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.Error_UnsupportedTrustTarget, trustTarget.ToString()));
+            }
+
+            if (trustTarget == VerificationTarget.Author && owners != null && owners.Any())
+            {
+                throw new ArgumentException(Strings.Error_TrustedAuthorNoOwners);
+            }
+
+            var v3ServiceIndex = string.Empty;
+            IRepositorySignature reposig = null;
+            var trustingRepository = trustTarget.HasFlag(VerificationTarget.Repository);
+
+            var primarySignature = await package.GetPrimarySignatureAsync(token);
+
+            if (primarySignature == null)
+            {
+                throw new InvalidOperationException(Strings.Error_PackageNotSigned);
+            }
+
+            if (trustingRepository)
+            {
+                if (primarySignature.Type == SignatureType.Repository)
+                {
+                    reposig = primarySignature as RepositoryPrimarySignature;
+                }
+                else
+                {
+                    var countersignature = RepositoryCountersignature.GetRepositoryCountersignature(primarySignature);
+                    reposig = countersignature ?? throw new InvalidOperationException(Strings.Error_RepoTrustExpectedRepoSignature);
+                }
+
+                v3ServiceIndex = reposig.V3ServiceIndexUrl.AbsoluteUri;
+            }
+
+            var signers = _trustedSignersProvider.GetTrustedSigners();
+            foreach (var existingSigner in signers)
+            {
+                if (string.Equals(existingSigner.Name, name, StringComparison.Ordinal))
+                {
+                    throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, Strings.Error_TrustedSignerAlreadyExists, name));
+                }
+
+                if (trustingRepository && existingSigner is RepositoryItem repoItem && string.Equals(repoItem.ServiceIndex, v3ServiceIndex, StringComparison.OrdinalIgnoreCase))
+                {
+                    throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, Strings.Error_TrustedRepoAlreadyExists, v3ServiceIndex));
+                }
+            }
+
+            if (trustingRepository)
+            {
+                var certificateItem = GetCertificateItemForSignature(reposig, allowUntrustedRoot);
+
+                _trustedSignersProvider.AddOrUpdateTrustedSigner(new RepositoryItem(name, v3ServiceIndex, CreateOwnersList(owners), certificateItem));
+            }
+
+            if (trustTarget.HasFlag(VerificationTarget.Author))
+            {
+                if (primarySignature.Type != SignatureType.Author)
```

Comment:
Shouldn't this check be before updating `_trustedSignersProvider`?

Reply:
It shouldn't matter, because the previous call we are doing to `AddOrUpdateTrustedSigner` is for adding a trusted repository, and this validation only affects adding a trusted author.